### PR TITLE
improvement(ort): restructering the OrtResultAnalyzer

### PR DESCRIPTION
**Pull request by neubs-bsi**
in preparation of integrating ORT and considering the sonarcloud result, I have restructured the OrtResultAnalyzer to be less ugly and complex.

**Done:**
- Introduct of OrtResultResolver abstract class and interface
- removing code from OrtResultAnalyzer to the resolver classe

Belongs to #22 and addition to #51

### CODE REVIEW

**Comment by sschuberth** [comment]
Commit message nit: "restructuring"

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtAnalyzerResultResolver.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package org.eclipse.sw360.antenna.ort.resolver;
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtAnalyzerResultResolver.java]: Missing license headers

Comment (neubs-bsi) [reply modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtAnalyzerResultResolver.java]: Very true... Sorry. I added it

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/workflow/analyzers/OrtResultAnalyzer.java [hunk 2]
```diff
@@ -111,209 +58,65 @@ public String getName() {
...
-    private Artifact mapArtifactFromAnalyzerResult(JsonNode ortPackage) {
-        Artifact artifact = new Artifact("OrtResult")
-                .addFact(mapFileName(ortPackage))
-                .addFact(new DeclaredLicenseInformation(mapDeclaredLicenses("declared_licenses", ortPackage)))
-                .addFact(new ArtifactSourceUrl(mapSourceUrl(ortPackage)));
-        mapCoordinates(ortPackage.get("id").textValue()).ifPresent(artifact::addFact);
-        mapHomepageUrl(ortPackage).ifPresent(homepageUrl -> artifact.addFact(new ArtifactHomepage(homepageUrl)));
+        scanResult.ifPresent(scnRslt -> {
+            OrtScannerResultResolver ortScannerResultResolver = new OrtScannerResultResolver(scnRslt);
+            Artifact artifact1 = ortScannerResultResolver.mapArtifact();
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/workflow/analyzers/OrtResultAnalyzer.java]: maybe better without the introduction of a temporary variable
```suggestion
            ortScannerResultResolver.mapArtifact()
                .askForAll(ArtifactFact.class)
                .forEach(artifact::addFact);
```

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/IOrtResultResolver.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import org.eclipse.sw360.antenna.model.artifact.facts.ArtifactCoordinates;
+import org.eclipse.sw360.antenna.model.artifact.facts.ArtifactSourceUrl;
+
+import java.util.Optional;
+
+public interface IOrtResultResolver {
+
+    Artifact mapArtifact();
+
+    Optional<ArtifactCoordinates> mapCoordinates(String ortIdentifier);
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/IOrtResultResolver.java]: not used externally, why is this part of the API

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/IOrtResultResolver.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    Optional<ArtifactSourceUrl> mapSourceUrl();
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/IOrtResultResolver.java]: Not used?

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/IOrtResultResolver.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.eclipse.sw360.antenna.ort.resolver;
+
+import org.eclipse.sw360.antenna.model.artifact.Artifact;
+import org.eclipse.sw360.antenna.model.artifact.facts.ArtifactCoordinates;
+import org.eclipse.sw360.antenna.model.artifact.facts.ArtifactSourceUrl;
+
+import java.util.Optional;
+
+public interface IOrtResultResolver {
```
[the remark this answers is not in the document]
Comment (maxhbr) [reply modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/IOrtResultResolver.java]: I think `IOrtResultResolver` could be replaced by `Function<JsonNode,Artifact>`. Or maybe even `Function<JsonNode,Collection<ArtifactFact>>` to get rid of the hack at https://github.com/eclipse/antenna/pull/199/files#diff-38c5935d20bf3705ad9bf2bd361f7490R115

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/workflow/analyzers/OrtResultAnalyzer.java [hunk 6]
```diff
@@ -111,209 +58,65 @@ public String getName() {
         JsonNode yamlContent = mapper.readTree(is);
         LOGGER.debug("Create artifact list from input stream");
 
-        List<Artifact> artifacts = new ArrayList<>();
-
-        boolean hasAnalyzerResults = !yamlContent.get("analyzer").isNull();
-        boolean hasScannerResults = !yamlContent.get("scanner").isNull();
+        Optional<JsonNode> optionalScanResults = !yamlContent.get("scanner").isNull() ?
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/workflow/analyzers/OrtResultAnalyzer.java]:
```suggestion
        Optional<JsonNode> optionalScanResults = Optional.ofNullable(yamlContent.get("scanner")).map(v -> v.get("results").get("scan_results"))
```

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtScannerResultResolver.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        return Optional.empty();
+    }
+
+    private Optional<ObservedLicenseInformation> mapObservedLicense() {
+        if(result.get("results").isArray()) {
+            Collection<String> licenses = new ArrayList<>();
+            Iterator<JsonNode> results = result.get("results").elements();
+            while(results.hasNext()) {
+                Iterator<JsonNode> findings = results.next().get("summary").get("license_findings").elements();
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtScannerResultResolver.java]: duplicated code, same line as in https://github.com/eclipse/antenna/pull/199/files#diff-5363d56ff7d691e5d496f9256fd61b73R82

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/AbstractOrtResultResolver.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import org.eclipse.sw360.antenna.model.artifact.facts.javaScript.JavaScriptCoordinates;
+import org.eclipse.sw360.antenna.model.xml.generated.MatchState;
+
+import java.util.Optional;
+
+public abstract class AbstractOrtResultResolver implements IOrtResultResolver{
+    JsonNode result;
+
+    public AbstractOrtResultResolver(JsonNode result) {
+        this.result = result;
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/AbstractOrtResultResolver.java]: I do not like the pattern implemented here, i.e.
```
    Artifact a = new AbstractOrtResultResolver(result).mapArtifact()
```
I would prefer the following alternative without a field variable to store the result:
```
    Artifact a = new AbstractOrtResultResolver().mapArtifact(result)
```

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtAnalyzerResultResolver.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            }
+            return Optional.of(new DeclaredLicenseInformation(LicenseSupport.mapLicenses(licenses)));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<ArtifactSourceUrl> mapSourceUrl() {
+        if (result.has("source_artifact")) {
+            return Optional.of(new ArtifactSourceUrl(result.get("source_artifact").get("url").textValue()));
```
Comment (maxhbr) [inline modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/OrtAnalyzerResultResolver.java]: this function (and all others with the `if (result.has(...)) { ... } return Optional.empty()` pattern) can be simplified to:
```java
Optioal.ofNullable(result.get("source_artifact"))
     .map(v -> v.get("url").textValue())
     .map(ArtifactSourceUrl::new);
```
see: https://fasterxml.github.io/jackson-databind/javadoc/2.8/com/fasterxml/jackson/databind/JsonNode.html#has(java.lang.String)

File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/AbstractOrtResultResolver.java (hunk 8 above)
Comment (blaumeiser-at-bosch) [reply modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/resolver/AbstractOrtResultResolver.java]: Ähem, new Abstract is a no go, either, something is abstract -> not instantiable, or it is not, than Abstract is not am option for the name.

**Pull request merged**